ambry-store/src/main/java/com.github.ambry.store/StatsManager.java

```diff
@@ -0,0 +1,202 @@
+      StatsManagerConfig config, Time time) throws IOException {
+    this.storageManager = storageManager;
+    initialPartitionIds = partitionIds;
+    statsOutputFile = new File(config.outputFilePath);
+    publishPeriodInSecs = config.publishPeriodInSecs;
+    metrics = new StatsManagerMetrics(registry);
+    this.time = time;
+  }
+
+  /**
+   * Start the stats manager by scheduling the periodic task that collect, aggregate and publish stats.
+   */
+  public void start() {
+    scheduler = Utils.newScheduler(1, false);
+    statsAggregator = new StatsAggregator();
+    // random initial delay between 1 to 10 minutes to offset nodes from collecting stats at the same time
+    scheduler.scheduleAtFixedRate(statsAggregator, new Random().nextInt(540) + 60, publishPeriodInSecs,
+        TimeUnit.SECONDS);
+  }
+
+  /**
+   * Stops the periodic task that is collecting, aggregating and publishing stats.
+   */
+  public void shutdown() throws InterruptedException {
+    if (statsAggregator != null) {
+      statsAggregator.cancel();
+    }
+    if (scheduler != null) {
+      scheduler.shutdown();
+      if (!scheduler.awaitTermination(30, TimeUnit.SECONDS)) {
+        logger.error("Could not terminate aggregator tasks after StatsManager shutdown");
+      }
+    }
+  }
+
+  /**
+   * Publishes stats to a local file in JSON format.
+   * @param statsWrapper the {@link StatsWrapper} to be published
+   * @throws IOException
+   */
+  void publish(StatsWrapper statsWrapper) throws IOException {
+    File tempFile = new File(statsOutputFile.getAbsolutePath() + ".tmp");
+    if (tempFile.createNewFile()) {
+      OutputStream outputStream = new BufferedOutputStream(new FileOutputStream(tempFile));
+      DatumWriter<StatsWrapper> statsWrapperDatumWriter = new SpecificDatumWriter<StatsWrapper>(StatsWrapper.class);
+      JsonEncoder jsonEncoder = EncoderFactory.get().jsonEncoder(StatsWrapper.getClassSchema(), outputStream, true);
+      try {
+        statsWrapperDatumWriter.write(statsWrapper, jsonEncoder);
+        jsonEncoder.flush();
+        outputStream.flush();
+      } finally {
+        outputStream.close();
+      }
+      if (!tempFile.renameTo(statsOutputFile)) {
+        throw new IOException(
+            "Failed to rename " + tempFile.getAbsolutePath() + " to " + statsOutputFile.getAbsolutePath());
+      }
+    } else {
+      throw new IOException("Temporary file creation failed when publishing stats " + tempFile.getAbsolutePath());
+    }
+  }
+
+  /**
+   * Fetch and aggregate stats from a given {@link Store}
+   * @param aggregatedSnapshot the {@link StatsSnapshot} to hold the aggregated result
+   * @param partitionId specifies the {@link Store} to be fetched from
+   * @param unreachableStores a {@link List} containing partition Ids that were unable to successfully fetch from
+   */
+  void collectAndAggregate(StatsSnapshot aggregatedSnapshot, PartitionId partitionId, List<String> unreachableStores) {
```

Comment:
do you think returning a Snapshot makes sense. Based on whether the arg is null or not, this method should generate one and return or add to the argument and return